Refactor and simplify structs (#2314)

keithchong opened the pull request:
The initial HAS struct types were maintained.  The goal now is to simplify, especially when the ComponentSpec configuration and the BindingComponentConfiguration have overlap. (They are both used primarily in the generation of the basic k8s deployment.yaml).  

Each element in the BindingComponentConfiguration is mapped nicely to the ComponentSpec, except for Resources, where in the former, the ResourceRequirement is a pointer, but, in the latter, it is a non-pointer. This PR will use a non-pointer.  Similarly, this reflects how it is done in k8s Container.  See
https://github.com/kubernetes/api/blob/6dd661f592205cc6035520138ebdbd271b9fa197/core/v1/types.go#L2369


The goal is ‘flatten’ the structs and simplify, so the elements from ComponentSpec will be moved to the Component struct.  The Source element will have the ComponentSource type however, and these types are flattened too. (There isn’t a need for multiple GitSources and Unions so this PR simplifies this.)

Mapping:
Original: component.Spec.Source.ComponentSourceUnion.GitSource, where GitSource is a pointer to GitSource
New: component.Source.ComponentSource, where Source is a pointer to ComponentSource

This is so that you can check for a nil ComponentSource.  See the corresponding changes in the Application Service PR. This is similar to checking if component.Spec.Source.ComponentSourceUnion.GitSource != nil 

Signed-off-by: Keith Chong <kykchong@redhat.com>

keithchong commented:
Changes to HAS are in this branch, excluding the go.mod file (just pull in the changes from this PR to see)

https://github.com/keithchong/application-service/tree/2314-ReactToRefactoredStructsFromGitOpsGenerator

keithchong commented:
Here is the commit (differences) in HAS:

https://github.com/redhat-appstudio/application-service/commit/e96d0e25f20b3112a6611e9a5ba7d0a0fb03cd08

kim-tsao commented on api/v1alpha1/component_types.go:
```diff
@@ -38,18 +39,11 @@ type GitSource struct {
 	DockerfileURL string `json:"dockerfileUrl,omitempty"`
```
Is there an order of precedence for component creation if the URL, DevfileURL, and DockerfileURLs are all specified at once?  Should this be documented?

keithchong replied:
Right now, they are placeholders.  Those are currently not used by the Generator - only the URL is used.

keithchong replied:
Based on your later review comments, I'll get rid of these.

kim-tsao replied:
I don't know if this will break HAS.  If these are HAS specific, then they belong in the HAS repo.  They don't really fit the category of "git source" either. cc @maysunfaisal

kim-tsao commented on pkg/gitops_test.go:
```diff
@@ -36,17 +36,11 @@ func TestCloneGenerateAndPush(t *testing.T) {
-			TargetPort: 5000,
+		ContainerImage: "testimage:latest",
+		Source: &gitopsv1alpha1.ComponentSource{
+			URL: repo,
```
Add coverage for ComponentSource optional fields

keithchong replied:
Will remove those fields.

kim-tsao commented on pkg/gitops_test.go:
```diff
@@ -1023,16 +1017,10 @@ func TestRemoveAndPush(t *testing.T) {
-			},
-			TargetPort: 5000,
+		Source: &gitopsv1alpha1.ComponentSource{
+			URL: repo,
```
Add coverage for ComponentSource optional fields

keithchong replied:
Removing those unused fields.

kim-tsao commented on pkg/gitops_test.go:
```diff
@@ -1558,20 +1546,14 @@ func TestGenerateAndPush(t *testing.T) {
-			TargetPort: 5000,
+		ContainerImage: "testimage:latest",
+		Source: &gitopsv1alpha1.ComponentSource{
+			URL: repo,
```
Add coverage for ComponentSource optional fields

keithchong replied:
Removing those unused fields.

maysunfaisal commented on api/v1alpha1/component_types.go:
```diff
@@ -38,18 +39,11 @@ type GitSource struct {
-	// Git Source for a Component
-	GitSource *GitSource `json:"git,omitempty"`
-}
+// Component is the Schema for the components API. This is used to generate any k8s deployment resource
```
I would generally tend to avoid discussing about any API that this library is not aware of, considering this is to be a standalone library

keithchong replied:
Changing it to:
```
// Component - An Application contains one or more components. This captures information about a component's
// deployment, service and route.
```

Although, the name of the struct, `Component` could also be changed.   eg. GeneratorOptions

maysunfaisal commented on api/v1alpha1/component_types.go:
```diff
@@ -38,18 +39,11 @@ type GitSource struct {
-}
+// Component is the Schema for the components API. This is used to generate any k8s deployment resource
+type Component struct {
+	// For Component.Name as an alternative to using ComponentName
+	metav1.ObjectMeta `json:"metadata,omitempty"`
```
Not sure why we want to introduce `metav1.ObjectMeta` here.. Any reason why Name is being replaced with this? ObjectMeta has more than Name that is perhaps not suited here

keithchong replied:
Pulled that out and will only use Name and Namespace

maysunfaisal commented on api/v1alpha1/component_types.go:
```diff
@@ -62,103 +56,27 @@ type ComponentSpec struct {
 
 	// Source describes the Component source
-	Source ComponentSource `json:"source,omitempty"`
+	Source *ComponentSource `json:"source,omitempty"`
```
how about 

```suggestion
	GitSource *GitSource `json:"gitSource,omitempty"`
```

because `ContainerImage` can also be a source.  This would mean also renaming `type ComponentSource struct{}`

I am thinking about making this change in the application-service too after thinking about it now. it looks cleaner

maysunfaisal commented on api/v1alpha1/component_types.go:
```diff
@@ -62,103 +56,27 @@ type ComponentSpec struct {
-	Conditions []metav1.Condition `json:"conditions,omitempty"`
+	// EnvironmentConfigEnvVar is an array of standard environment variables in addition to the component EnvVars.
+	// These will ONLY be added to the deployment patches overlays deployment.yaml
+	EnvironmentConfigEnvVar []corev1.EnvVar `json:"env"`
```
what is this replacing? I am confused here

Also the json tag does not match the name, so the linter is throwing a warning here

maysunfaisal replied:
OK i see what this is trying to achieve. I think this will be confusing to leave it on the same level..

How about introducing an intermediate level here called Environment?

so component.environment.environmentconfig is semantically more easier to understand.. WDYT

keithchong replied:
Using BaseEnvVar and OverlayEnvVar

keithchong commented on pkg/gitops.go:
```diff
@@ -107,14 +107,14 @@ func CommitAndPush(outputPath string, repoPathOverride string, remote string, co
 func GenerateAndPush(outputPath string, remote string, component gitopsv1alpha1.Component, e Executor, appFs afero.Afero, branch string, doPush bool, createdBy string, commonStorage *corev1.PersistentVolumeClaim) error {
```
Done

keithchong commented:
Note that the definition of Resource is different in GitOps service and HAS:

Managed GitOps Services

https://github.com/redhat-appstudio/managed-gitops/blob/864c0c9c5a743275a4b0e497654a9220a4fb993f/appstudio-shared/apis/appstudio.redhat.com/v1alpha1/applicationsnapshotenvironmentbinding_types.go#L67

HAS

https://github.com/redhat-appstudio/application-service/blob/0b7dd2d0dc4d9a31c6803a9bbbab2b985ddbbcdf/api/v1alpha1/component_types.go#L89

Both are used when generating the deployment file (deployment kind).  I've changed it to be a non-pointer.  See the component_types.go file.

keithchong commented:
Here is the updated commit in HAS to reflect these changes

https://github.com/redhat-appstudio/application-service/commit/93fd73aca2960d5789bfaee3006067409e13ac13

keithchong commented:
Here is the latest commit to the HAS repo to reflect these changes.  A follow-on PR for that will be created including updating the go.mod file, etc.  (It just has the source changes)

https://github.com/redhat-appstudio/application-service/commit/a951ab06c36f8a9ca081eb63249858d0addb3f1a

maysunfaisal commented on api/v1alpha1/component_types.go:
```diff
@@ -18,147 +18,70 @@ package v1alpha1
 
 	// Specify a branch/tag/commit id. If not specified, default is `main`/`master`.
-	Revision string `json:"revision,omitempty"`
+	//Revision string `json:"revision,omitempty"`
```
I think we might need to keep this

keithchong replied:
These are not used anywhere in the source so far.

kim-tsao replied:
It is used in the `application-service` repo which will eventually rely on this code (see https://github.com/redhat-appstudio/application-service/blob/651fc2e107c413f88dc3796b581ef13e3387243c/controllers/component_controller.go#L210)

keithchong replied:
That is application-service's original types. Note that a duplicate of these types (structs) were made in the common library to make the transition to the library easier. This duplicate or the 'mapping' now needs to be reduced because many of the fields are not necessary when generating the k8s resources.   If someone you need any field added to the 'mapping', then that field must be used by the library in some way.  Currently, what you see now in this PR are the fields that are used.

maysunfaisal commented on pkg/generate.go:
```diff
@@ -77,7 +77,7 @@ func Generate(fs afero.Afero, gitOpsFolder string, outputFolder string, componen
 
 // GenerateOverlays generates the overlays dir from a given BindingComponent
-func GenerateOverlays(fs afero.Afero, gitOpsFolder string, outputFolder string, component gitopsv1alpha1.BindingComponentConfiguration, environment gitopsv1alpha1.Environment, imageName, namespace string, componentGeneratedResources map[string][]string) error {
+func GenerateOverlays(fs afero.Afero, gitOpsFolder string, outputFolder string, component gitopsv1alpha1.GeneratorOptions, imageName, namespace string, componentGeneratedResources map[string][]string) error {
```
```suggestion
func GenerateOverlays(fs afero.Afero, gitOpsFolder string, outputFolder string, options gitopsv1alpha1.GeneratorOptions, imageName, namespace string, componentGeneratedResources map[string][]string) error {
```

cleaner

keithchong replied:
Done

maysunfaisal commented on pkg/generate.go:
```diff
@@ -258,7 +258,7 @@ func generateDeployment(component gitopsv1alpha1.Component) *appsv1.Deployment {
 }
 
-func generateDeploymentPatch(component gitopsv1alpha1.BindingComponentConfiguration, environment gitopsv1alpha1.Environment, imageName, namespace string) *appsv1.Deployment {
+func generateDeploymentPatch(component gitopsv1alpha1.GeneratorOptions, imageName, namespace string) *appsv1.Deployment {
```
```suggestion
func generateDeploymentPatch(options gitopsv1alpha1.GeneratorOptions, imageName, namespace string) *appsv1.Deployment {
```

keithchong replied:
Done

maysunfaisal commented on pkg/generate.go:
```diff
@@ -284,15 +284,15 @@ func generateDeploymentPatch(component gitopsv1alpha1.BindingComponentConfigurat
 	}
 
-	for _, env := range component.Env {
+	for _, env := range component.BaseEnvVar {
```
```suggestion
	for _, env := range options.OverlayEnvVar {
```

since this is in patch

maysunfaisal replied:
Is my suggestion correct?

keithchong replied:
`component.Env` was the original env var.  We're no just renaming it to `BaseEnvVar`.

maysunfaisal commented on pkg/generate.go:
```diff
@@ -314,14 +314,12 @@ func generateDeploymentPatch(component gitopsv1alpha1.BindingComponentConfigurat
 }
 
-func generateService(component gitopsv1alpha1.Component) *corev1.Service {
+func generateService(component gitopsv1alpha1.GeneratorOptions) *corev1.Service {
```
```suggestion
func generateService(options gitopsv1alpha1.GeneratorOptions) *corev1.Service {
```

keithchong replied:
Done

maysunfaisal commented on pkg/generate.go:
```diff
@@ -348,7 +346,7 @@ func generateService(component gitopsv1alpha1.Component) *corev1.Service {
 }
 
-func generateRoute(component gitopsv1alpha1.Component) *routev1.Route {
+func generateRoute(component gitopsv1alpha1.GeneratorOptions) *routev1.Route {
```
```suggestion
func generateRoute(options gitopsv1alpha1.GeneratorOptions) *routev1.Route {
```

keithchong replied:
Done

maysunfaisal commented on pkg/generate.go:
```diff
@@ -388,9 +386,9 @@ func generateRoute(component gitopsv1alpha1.Component) *routev1.Route {
-func getReplicas(component gitopsv1alpha1.Component) int32 {
-	if component.Spec.Replicas > 0 {
-		return int32(component.Spec.Replicas)
+func getReplicas(component gitopsv1alpha1.GeneratorOptions) int32 {
```
```suggestion
func getReplicas(options gitopsv1alpha1.GeneratorOptions) int32 {
```

keithchong replied:
Done

maysunfaisal commented on pkg/generate.go:
```diff
@@ -401,19 +399,22 @@ func getReplicas(component gitopsv1alpha1.Component) int32 {
 // app.kubernetes.io/managed-by: "kustomize"
 // app.kubernetes.io/created-by: "application-service"
-func generateK8sLabels(component gitopsv1alpha1.Component) map[string]string {
+func generateK8sLabels(component gitopsv1alpha1.GeneratorOptions) map[string]string {
```
```suggestion
func generateK8sLabels(options gitopsv1alpha1.GeneratorOptions) map[string]string {
```

keithchong replied:
Done

maysunfaisal commented on pkg/generate.go:
```diff
@@ -401,19 +399,22 @@ func getReplicas(component gitopsv1alpha1.Component) int32 {
 // GetMatchLabel returns the label selector that will be used to tie deployments, services, and pods together
 // For cleanliness, using just one unique label from the generateK8sLabels function
-func getMatchLabel(component gitopsv1alpha1.Component) map[string]string {
+func getMatchLabel(component gitopsv1alpha1.GeneratorOptions) map[string]string {
```
```suggestion
func getMatchLabel(options gitopsv1alpha1.GeneratorOptions) map[string]string {
```

maysunfaisal commented on pkg/gitops.go:
```diff
@@ -48,7 +47,7 @@ type Executor interface {
 // 8. The gitops config containing the build bundle;
 // Adapted from https://github.com/redhat-developer/kam/blob/master/pkg/pipelines/utils.go#L79
-func CloneGenerateAndPush(outputPath string, remote string, component gitopsv1alpha1.Component, e Executor, appFs afero.Afero, branch string, context string, doPush bool) error {
+func CloneGenerateAndPush(outputPath string, remote string, component gitopsv1alpha1.GeneratorOptions, e Executor, appFs afero.Afero, branch string, context string, doPush bool) error {
```
```suggestion
func CloneGenerateAndPush(outputPath string, remote string, options gitopsv1alpha1.GeneratorOptions, e Executor, appFs afero.Afero, branch string, context string, doPush bool) error {
```

keithchong replied:
Done

maysunfaisal commented on pkg/gitops.go:
```diff
@@ -104,17 +111,27 @@ func CommitAndPush(outputPath string, repoPathOverride string, remote string, co
+// 6. The branch to push to
+// 7. Optionally push to the GitOps repository or not.  Default is not to push.
+// 8. createdBy: Use a unique name to identify that clients are generating the GitOps repository. Default is "application-service" and should be overwritten.
+func GenerateAndPush(outputPath string, remote string, component gitopsv1alpha1.GeneratorOptions, e Executor, appFs afero.Afero, branch string, doPush bool, createdBy string) error {
```
```suggestion
func GenerateAndPush(outputPath string, remote string, options gitopsv1alpha1.GeneratorOptions, e Executor, appFs afero.Afero, branch string, doPush bool, createdBy string) error {
```

keithchong replied:
Done

maysunfaisal commented on pkg/gitops.go:
```diff
@@ -104,17 +111,27 @@ func CommitAndPush(outputPath string, repoPathOverride string, remote string, co
-	gitHostAccessToken := component.Spec.Secret
-	gitOpsRepoURL := component.Spec.Source.GitSource.URL
+	gitHostAccessToken := component.Secret
+	gitOpsRepoURL := component.GitSource.URL
```
`component.GitSource` will panic if its nil

keithchong replied:
Thanks for catching this.  I also moved it into the `if doPush` block just below where it is actually used.

maysunfaisal commented on pkg/gitops.go:
```diff
@@ -190,7 +207,21 @@ func GenerateAndPush(outputPath string, remote string, component gitopsv1alpha1.
+// 12. The path within the repository to generate the resources in
+// 13. Push the changes to the repository or not.
+// 14. The gitops config containing the build bundle;
+func GenerateOverlaysAndPush(outputPath string, clone bool, remote string, component gitopsv1alpha1.GeneratorOptions, applicationName, environmentName, imageName, namespace string, e Executor, appFs afero.Afero, branch string, context string, doPush bool, componentGeneratedResources map[string][]string) error {
```
```suggestion
func GenerateOverlaysAndPush(outputPath string, clone bool, remote string, options gitopsv1alpha1.GeneratorOptions, applicationName, environmentName, imageName, namespace string, e Executor, appFs afero.Afero, branch string, context string, doPush bool, componentGeneratedResources map[string][]string) error {
```

keithchong replied:
Done

maysunfaisal reviewed (commented):
- Just a suggestion to change function param to `options` since we changed the struct name
- pls have a look at my ENV suggestion for deployment patch func, is it valid?

keithchong commented:
> * Just a suggestion to change function param to `options` since we changed the struct name
> * pls have a look at my ENV suggestion for deployment patch func, is it valid?

I also renamed the file in api/v1alpha1 from `component_types.go` to `generator_options.go`

maysunfaisal commented on api/v1alpha1/generator_options.go:
```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2021 Red Hat, Inc.
```
Also while we are at it, could you do a Cmd-F and search if the copyright years are correct and update it

keithchong replied:
Done.  It was just that file that needed to be updated

keithchong commented:
Here are the latest rebased changes to the application-service to react to these changes (minus the go.mod changes)

https://github.com/redhat-appstudio/application-service/commit/94e3c9907510bf8fb350b310e89dfa7edb549754

keithchong commented:
Thanks for the review folks.  We needed to rebase again.
